enhancements/etcd/scaling-etcd-with-raft-learners.md

```diff
@@ -0,0 +1,343 @@
+---
+title: Scaling etcd with Raft Learners
+authors:
+- "@hexfusion"
+  reviewers:
+- "@deads2k"
+- "@lilic"
+- "@hasbro17"
+- "@joelspeed"
+- "@jeremyeder"
+  approvers:
+- "@mfojtik"
+  creation-date: 2021-10-04
+  last-updated: 2021-10-04
+  status: implementable
+  see-also: []
+  replaces: []
+  superseded-by: []
+---
+
+# Scaling etcd with Raft Learners
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Over time as clusters live longer and workloads grow the ability to scale the control-plane and replace failed nodes
+becomes a critical part of the admins maintenance overhead. Today the `cluster-etcd-operator` manages scaling up of
+the etcd cluster. To provide the foundation for initiatives such as scale down and vertical control plane scaling[1].
+The `cluster-etcd-operator` must ensure proper safety mechanisms exist to adjust membership of the etcd cluster.
+
+Introduced in etcd v3.4 the raft learner[2] provides mitigations which reduces quorum and
+stability issues during scaling. A learner is essentially an etcd member which is non voting thus can not impact 
+quorum but like other members will receive log replications from the leader.
+
+This enhancement proposes:
+- Replacing the default scale up performed by the cluster-etcd-operator to use raft learners.
+- Deprecation and removal of the current `discover-etcd-initial-cluster`[3] command and replacing its
+functionality with the existing `etcdenvvar` controller and subcommands of the `cluster-etcd-operator`. By
+eliminating this code it also removes the teams largest carry against upstream etcd.
+- Ensuring that scale up only happens on control-plane nodes which have a quorum-guard pod monitoring health.
+- Adding a flag to etcd which allows configuration of maximum learners in cluster (--max-learners). Today the max is 1.
```

Comment:
So, I'm wondering what happens in a scenario where two new "masters" are added to the cluster at once. Say we have a healthy 3 node control plane, 2 new etcd static pods start. They now both want to become learners, but only one is allowed (due to the max-learners flag). What happens to the one that isn't allowed to be a learner?

Reply:
@JoelSpeed I'm guessing first one past the post wins/gets added as a learner:
https://etcd.io/docs/v3.4/op-guide/runtime-configuration/#error-cases-when-adding-a-learner-member
For the other one I imagine the etcd-operator would probably keep retrying to add it as a learner and failing to do so.

But to your point of 2 static pods, do you mean on a single node? I'm wondering how that scenario might arise.

Since the targetconfigcontroller is just [applying static pod configurations](https://github.com/openshift/cluster-etcd-operator/blob/10c3297019a692f714b7766872b26c67238038d9/pkg/operator/targetconfigcontroller/targetconfigcontroller.go#L194-L203) via the configmap, I'm assuming the library-go static pod controller would just create a single manifest in the static pod directory for that node.
But if something goes wrong and we do somehow end up with two identical etcd static pods on a single node, I'm not sure if the etcd-operator can fix that situation by just removing that extra member or try to directly update static pod manifests on the node. That's probably a library-go issue at that point right @hexfusion ?

Although we probably should account for that possibility by recognizing that we've reached the max number of learner members from the error returned.

Reply:
This is a good question I will add an addition of a matrix that describes these conditions and expected results.

Reply:
> We'll probably get an alert if the 2nd etcd pod doesn't get added as a learner member after some time, and if the 1st one that is a learner doesn't get promoted in time to make room for the 2nd one.

Probably is good, definite is better 😉  We should check if these alerts exist, check whether that's how we would want to report this, and add the new alerts/make a note they exist already to the proposal as appropriate 😄

Reply:
If 2 members were added to an existing 3 member cluster wouldn't it be intentional?  Can the number of expected learners be changed dynamically and what is the impact of that change?  If so, then managing max learners is an operational concern and blocking additional learners and raising an alert is the expected behavior.